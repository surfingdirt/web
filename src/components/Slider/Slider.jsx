import PropTypes from 'prop-types';
import React, { Fragment, PureComponent } from 'react';
import { Carousel } from 'react-responsive-carousel';

import 'react-responsive-carousel/lib/styles/carousel.css';

export default class Slider extends PureComponent {
  static propTypes = {
    children: PropTypes.node.isRequired,
  };

  render() {
    // See http://react-responsive-carousel.js.org/storybook/ for supported settings
    const { children, ...settings } = this.props;

    return (
      <Fragment>
        <Carousel {...settings} useKeyboardArrows>
          {children}
        </Carousel>
      </Fragment>
    );
  }
}
