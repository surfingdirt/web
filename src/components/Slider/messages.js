const messages = (getText, getPlural) => (key) =>
  ({
    previous: getText(
      'Previous',
      'Slider',
    ) /* Label of a button to scroll a list of items back */,
    next: getText(
      'Next',
      'Slider',
    ) /* Label of a button to scroll a list of items forward */,
  }[key]);

export default messages;
