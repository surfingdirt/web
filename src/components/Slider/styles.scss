@import "../../styles/colors";
@import "../../styles/constants";

.container {
  display: flex;
}

.slideContainer {
  display: flex;
  width: 100%;
  white-space: nowrap;
  &.withTransitions {
    transition: all .5s;
  }
}

.sliderOverflow {
  height: 100%;
  width: 100%;
  overflow: hidden;
}
