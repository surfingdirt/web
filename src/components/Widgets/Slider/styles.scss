@import "../../../styles/colors";
@import "../../../styles/constants";
@import "../../../styles/media-queries";

.container {
  display: flex;
}

.slideContainer {
  display: flex;
  width: 100%;
  white-space: nowrap;
  -webkit-overflow-scrolling: touch;
  overflow-x: scroll;

  @media(min-width: $desktop-layout-threshold-min) {
    overflow-x: visible;
  }

  &.withTransitions {
    transition: all .5s;
  }
}

.sliderOverflow {
  height: 100%;
  width: 100%;
  overflow: hidden;
}
