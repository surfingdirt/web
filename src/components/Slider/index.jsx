import React from 'react';
import classNames from 'classnames';
import PropTypes from 'prop-types';

import Translate from 'Hocs/Translate';
import icons, { getIcon } from 'Utils/icons';
import AppContext from '~/contexts';

import messages from './messages';
import styles from './styles.scss';

const { NEXT, PREVIOUS } = icons;

const RTL = 'rtl';
const FORWARD = 'forward';
const BACK = 'back';

const INITIAL_STATE = {
  targetIndex: 0,
  targetPosition: 0,
  touching: false,
  mounted: false,
  isAtBeginning: true,
  isAtEnd: false,
  overflows: false,
};
const INITIAL_QUICK_STATE = {
  swipeTranslateX: null,
  touchStartX: null,
  touchMoveX: null,
};

const clamp = (min, value, max) => {
  if (value < min) {
    return min;
  }
  if (value > max) {
    return max;
  }
  return value;
};

class Slider extends React.PureComponent {
  static contextType = AppContext;

  static propTypes = {
    children: PropTypes.node.isRequired,
    t: PropTypes.func.isRequired,
    className: PropTypes.string,
    nextClassName: PropTypes.string,
    onSlideChange: PropTypes.func,
    prevClassName: PropTypes.string,
  };

  static defaultProps = {
    className: null,
    nextClassName: null,
    onSlideChange: () => null,
    prevClassName: null,
  };

  constructor(props) {
    super(props);

    this.stepRefs = [];
    this.state = Object.assign({}, INITIAL_STATE);
    this.quickState = Object.assign({}, INITIAL_QUICK_STATE);

    this.performSlide = this.performSlide.bind(this);
    this.touchStart = this.touchStart.bind(this);
    this.touchMove = this.touchMove.bind(this);
    this.touchEnd = this.touchEnd.bind(this);
  }

  componentDidMount() {
    const { scrollWidth, offsetWidth } = this.sliderEl;
    const overflows = scrollWidth > offsetWidth;

    this.setState({
      mounted: true,
      overflows,
    });
  }

  getTargetPosition(index) {
    const target = this.stepRefs[index];
    return Math.max(0, target.offsetLeft - this.stepRefs[0].offsetLeft);
  }

  setQuickState(updates) {
    this.quickState = Object.assign(this.quickState, updates);
  }

  updateSliderPosition(position) {
    this.sliderEl.style.transform = `translate(${-position}px)`;
  }

  touchStart(e) {
    const {
      state: { targetPosition },
    } = this;
    const { scrollWidth, offsetWidth } = this.sliderEl;
    const maxSwipe = scrollWidth - offsetWidth;
    const { clientX } = e.touches[0];
    this.setQuickState({
      touchStartX: clientX,
      touchMoveX: clientX,
      maxSwipe,
      currentSwipePosition: targetPosition,
    });
    this.setState({ touching: true });
  }

  touchMove(e) {
    return;
    const touchMoveX = e.touches[0].clientX;
    const {
      state: { targetPosition },
      quickState: { touchStartX, maxSwipe },
    } = this;

    const deltaX = touchStartX - touchMoveX;

    let currentSwipePosition = targetPosition + deltaX;
    currentSwipePosition = clamp(0, currentSwipePosition, maxSwipe);

    this.setQuickState({ touchMoveX, currentSwipePosition });
    this.updateSliderPosition(currentSwipePosition);
  }

  touchEnd() {
    return;
    const {
      props: { onSlideChange, children },
      state: { targetIndex, targetPosition },
      quickState: { currentSwipePosition },
    } = this;
    // compute new slide index
    // init search and result
    let minDist = Number.MAX_SAFE_INTEGER;
    let newTargetIndex = 0;
    let nextPos = 0;
    // search forward or backward from current target
    const shouldIncrement = targetPosition < currentSwipePosition;

    let idx = targetIndex;
    while (idx >= 0 && idx <= children.length - 1) {
      const elementPosition = this.getTargetPosition(idx);
      const dist = Math.abs(elementPosition - currentSwipePosition);
      if (minDist < dist) {
        // we are not improving anymore, break
        break;
      }
      minDist = dist;
      newTargetIndex = idx;
      nextPos = elementPosition;

      idx += shouldIncrement ? 1 : -1;
    }
    this.updateSliderPosition(nextPos);
    // reset
    this.setQuickState({ currentSwipePosition: null, touchStartX: null, touchMoveX: null });
    this.setState({ targetIndex: newTargetIndex, targetPosition: nextPos, touching: false }, () => {
      onSlideChange(newTargetIndex);
    });
  }

  async performSlide(slideDir) {
    const {
      state: { targetIndex, targetPosition: oldTargetPosition },
      props: { onSlideChange },
      context: { dir },
    } = this;

    const { scrollWidth, offsetWidth } = this.sliderEl;
    const maxTranslateX = scrollWidth - offsetWidth;

    const increment = (slideDir === FORWARD ? 1 : -1) * (dir === RTL ? -1 : 1);
    let newTargetIndex = targetIndex + increment;
    newTargetIndex = clamp(0, newTargetIndex, this.stepRefs.length - 1);

    const targetPosition = clamp(0, this.getTargetPosition(newTargetIndex), maxTranslateX);
    const isAtEnd = targetPosition >= maxTranslateX;
    const isAtBeginning = targetPosition === 0;

    if (oldTargetPosition === targetPosition) {
      return;
    }
    this.setState(
      {
        targetIndex: newTargetIndex,
        targetPosition,
        isAtBeginning,
        isAtEnd,
      },
      () => {
        onSlideChange(newTargetIndex);
      },
    );
  }

  render() {
    const {
      props: { children, t, className, nextClassName, prevClassName },
      state: { targetPosition, touching, mounted, isAtBeginning, isAtEnd, overflows },
      quickState: { currentSwipePosition },
      stepRefs,
    } = this;
    const sliderStyle = {};

    if (mounted && children.length > 0) {
      if (touching) {
        sliderStyle.transform = `translate(${-currentSwipePosition}px)`;
      } else {
        sliderStyle.transform = `translate(${-targetPosition}px)`;
      }
    }
    return (
      <div className={classNames(styles.container, className)}>
        <div className={styles.sliderOverflow}>
          <div
            ref={(ref) => {
              this.sliderEl = ref;
            }}
            onTouchStart={this.touchStart}
            onTouchMove={this.touchMove}
            onTouchEnd={this.touchEnd}
            style={sliderStyle}
            className={classNames(styles.slideContainer, !touching && styles.withTransitions)}
          >
            {React.Children.map(children, (child, index) =>
              React.cloneElement(child, { ref: (inst) => (stepRefs[index] = inst) }),
            )}
          </div>
        </div>
        {!isAtBeginning && overflows && (
          <button
            className={classNames('rtlTransform', prevClassName)}
            type="button"
            aria-label={t('previous')}
            onClick={() => this.performSlide(BACK)}
          >
            {getIcon({ type: PREVIOUS })}
          </button>
        )}
        {!isAtEnd && overflows && (
          <button
            className={classNames('rtlTransform', nextClassName)}
            type="button"
            aria-label={t('next')}
            onClick={() => this.performSlide(FORWARD)}
          >
            {getIcon({ type: NEXT })}
          </button>
        )}
      </div>
    );
  }
}

export default Translate(messages)(Slider);
