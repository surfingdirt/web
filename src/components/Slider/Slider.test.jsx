import React, { Fragment } from 'react';

import Slider from './Slider';

const children = (
  <Fragment>
    <div>
      <img src="#" alt="Legend 1" />
      <p className="legend">Legend 1</p>
    </div>
    <div>
      <img src="#" alt="Legend 2" />
      <p className="legend">Legend 2</p>
    </div>
  </Fragment>
);

describe('Slider', () => {
  it('should call a Carousel with default options', () => {
    const wrapper = shallow(<Slider showThumbs={false}>{children}</Slider>);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render children wrapped in its own markup', () => {
    const wrapper = mount(<Slider showThumbs={false}>{children}</Slider>);
    expect(wrapper).toMatchSnapshot();

    const carouselRoot = wrapper.find('.carousel-slider');
    expect(carouselRoot).toHaveLength(1);

    const items = carouselRoot.find('li');
    expect(items).toHaveLength(2);

    expect(items.find('.legend')).toHaveLength(2);
    expect(
      items
        .find('.legend')
        .at(0)
        .text(),
    ).toEqual('Legend 1');
    expect(
      items
        .find('.legend')
        .at(1)
        .text(),
    ).toEqual('Legend 2');
  });
});
