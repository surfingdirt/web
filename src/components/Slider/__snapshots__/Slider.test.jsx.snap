// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider should call a Carousel with default options 1`] = `
<Fragment>
  <Carousel
    autoPlay={false}
    axis="horizontal"
    centerMode={false}
    centerSlidePercentage={80}
    dynamicHeight={false}
    emulateTouch={false}
    infiniteLoop={false}
    interval={3000}
    onChange={[Function]}
    onClickItem={[Function]}
    onClickThumb={[Function]}
    selectedItem={0}
    showArrows={true}
    showIndicators={true}
    showStatus={true}
    showThumbs={false}
    statusFormatter={[Function]}
    stopOnHover={true}
    swipeScrollTolerance={5}
    swipeable={true}
    transitionTime={350}
    useKeyboardArrows={true}
    verticalSwipe="standard"
    width="100%"
  >
    <div>
      <img
        alt="Legend 1"
        src="#"
      />
      <p
        className="legend"
      >
        Legend 1
      </p>
    </div>
    <div>
      <img
        alt="Legend 2"
        src="#"
      />
      <p
        className="legend"
      >
        Legend 2
      </p>
    </div>
  </Carousel>
</Fragment>
`;

exports[`Slider should render children wrapped in its own markup 1`] = `
<div>
  <div
    className="carousel carousel-slider"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <button
      className="control-arrow control-prev control-disabled"
      onClick={[Function]}
      type="button"
    />
    <div
      className="slider-wrapper axis-horizontal"
      style={Object {}}
    >
      <ul
        className="slider animated"
        onMouseDown={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "MozTransform": "translate3d(0%,0,0)",
            "MozTransitionDuration": "350ms",
            "MsTransform": "translate3d(0%,0,0)",
            "MsTransitionDuration": "350ms",
            "OTransform": "translate3d(0%,0,0)",
            "OTransitionDuration": "350ms",
            "WebkitTransform": "translate3d(0%,0,0)",
            "WebkitTransitionDuration": "350ms",
            "msTransform": "translate3d(0%,0,0)",
            "msTransitionDuration": "350ms",
            "transform": "translate3d(0%,0,0)",
            "transitionDuration": "350ms",
          }
        }
      >
        <li
          className="slide selected"
          key="itemKey0/.0"
          onClick={[Function]}
        >
          <div>
            <img
              alt="Legend 1"
              src="#"
            />
            <p
              className="legend"
            >
              Legend 1
            </p>
          </div>
          <div>
            <img
              alt="Legend 2"
              src="#"
            />
            <p
              className="legend"
            >
              Legend 2
            </p>
          </div>
        </li>
      </ul>
    </div>
    <button
      className="control-arrow control-next control-disabled"
      onClick={[Function]}
      type="button"
    />
    <ul
      className="control-dots"
    >
      <li
        className="dot selected"
        key="0/.0"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
        value={0}
      />
    </ul>
    <p
      className="carousel-status"
    >
      1 of 1
    </p>
  </div>
</div>
`;
